api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+		return err
+	}
+
+	// choose which instances to terminate during our scale-down process
+	// instead of having asg randomly select instances
+	// e.g. if we scale from 5 to 3, we can terminate up to 2 unused instances
+	currentCapacity := int(aws.Int64Value(asg.DesiredCapacity))
+	maxNumberOfInstancesToTerminate := currentCapacity - desiredScale
+
+	canTerminate := func(i int) bool {
+		if i+1 > maxNumberOfInstancesToTerminate {
+			return false
+		}
+
+		if i > len(unusedProviders)-1 {
+			return false
+		}
+
+		return true
+	}
+
+	for i := 0; canTerminate(i); i++ {
+		unusedProvider := unusedProviders[i]
+		log.Printf("[DEBUG] [EnvironmentScaler] Terminating unused instance '%s' from environment '%s'.", unusedProvider.ID, clusterName)
+
+		input := &autoscaling.TerminateInstanceInAutoScalingGroupInput{}
+		input.SetInstanceId(unusedProvider.ID)
+		input.SetShouldDecrementDesiredCapacity(false)
+
+		if _, err := e.Client.AutoScaling.TerminateInstanceInAutoScalingGroup(input); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *EnvironmentScaler) scaleTo(clusterName string, desiredScale int, unusedProviders []*ResourceProvider) error {
+	input := &autoscaling.DescribeAutoScalingGroupsInput{}
+	input.SetAutoScalingGroupNames([]*string{&clusterName})
+
+	asgs, err := e.Client.AutoScaling.DescribeAutoScalingGroups(input)
+	if err != nil {
+		return err
+	}
+
+	asg := asgs.AutoScalingGroups[0]
+
+	currentCapacity := int(aws.Int64Value(asg.DesiredCapacity))
+
+	switch {
+	case desiredScale > currentCapacity:
+		log.Printf("[DEBUG] [EnvironmentScaler] Attempting to scale environment '%s' from current scale of '%d' to desired scale of '%d'.", clusterName, currentCapacity, desiredScale)
+		return e.scaleUp(clusterName, desiredScale, asg)
+
+	case desiredScale < currentCapacity:
+		log.Printf("[DEBUG] [EnvironmentScaler] Attempting to scale environment '%s' from current scale of '%d' to desired scale of '%d'.", clusterName, currentCapacity, desiredScale)
+		return e.scaleDown(clusterName, desiredScale, asg, unusedProviders)
+
+	default:
+		log.Printf("[DEBUG] [EnvironmentScaler] Environment '%s' is at desired scale of '%d'. No scaling action required.", clusterName, currentCapacity)
+		return nil
+	}
+}
+
+func (e *EnvironmentScaler) scaleUp(clusterName string, desiredScale int, asg *autoscaling.Group) error {
+	maxCapacity := int(aws.Int64Value(asg.MaxSize))
+	if desiredScale > maxCapacity {
+		log.Printf("[DEBUG] [EnvironmentScaler] Maximum capacity is '%d'. Aborting scaling action for environment '%s'.", maxCapacity, clusterName)
+		return nil
+	}
+
+	input := &autoscaling.SetDesiredCapacityInput{}
+	input.SetAutoScalingGroupName(aws.StringValue(asg.AutoScalingGroupName))
+	input.SetDesiredCapacity(int64(desiredScale))
+
+	if _, err := e.Client.AutoScaling.SetDesiredCapacity(input); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func sortConsumersByCPU(c []ResourceConsumer) {
```

Comment:
I don't think this ever be needed.

Reply:
do consumers ever get sorted?

Reply:
They do, near the top of `calculateScaleUp` - this was a result of our conversation about sorting both the "cups" and the "rocks".